// @matterlabs/hardhat-zksync-solc/dist/src/compile/docker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compileWithDocker = exports.pullImageIfNecessary = exports.createDocker = exports.validateDockerIsInstalled = exports.dockerImage = void 0;
const hardhat_docker_1 = require("@nomiclabs/hardhat-docker");
const utils_1 = require("../utils");
const stream_1 = require("stream");
async function runZksolcContainer(docker, image, command, input) {
    const createOptions = {
        Tty: false,
        AttachStdin: true,
        OpenStdin: true,
        StdinOnce: true,
        HostConfig: {
            AutoRemove: true,
        },
        Cmd: command,
        Image: hardhat_docker_1.HardhatDocker.imageToRepoTag(image),
    };
    const container = await docker.createContainer(createOptions);
    let output = Buffer.from('');
    let chunk = Buffer.from('');
    const stream = new stream_1.Writable({
        write: function (incoming, _encoding, next) {
            // Please refer to the 'Stream format' chapter at
            // https://docs.docker.com/engine/api/v1.37/#operation/ContainerAttach
            // to understand the details of this implementation.
            chunk = Buffer.concat([chunk, incoming]);
            let size = chunk.readUInt32BE(4);
            while (chunk.byteLength >= 8 + size) {
                output = Buffer.concat([output, chunk.slice(8, 8 + size)]);
                chunk = chunk.slice(8 + size);
                if (chunk.byteLength >= 8) {
                    size = chunk.readUInt32BE(4);
                }
            }
            next();
        },
    });
    const dockerStream = await container.attach({
        stream: true,
        stdin: true,
        stdout: true,
        stderr: true,
        hijack: true,
    });
    dockerStream.pipe(stream);
    await container.start();
    dockerStream.end(input);
    await container.wait();
    const compilerOutput = output.toString('utf8');
    try {
        return JSON.parse(compilerOutput);
    }
    catch (_a) {
        throw (0, utils_1.pluginError)(compilerOutput);
    }
}
// Notice: contents of this file were mostly copy-pasted from the official Hardhat Vyper plugin
// https://github.com/nomiclabs/hardhat/tree/master/packages/hardhat-vyper
function dockerImage(imageName) {
    if (!imageName) {
        throw (0, utils_1.pluginError)('Docker source was chosen but no image was specified');
    }
    return {
        repository: imageName,
        tag: 'latest',
    };
}
exports.dockerImage = dockerImage;
async function validateDockerIsInstalled() {
    if (!(await hardhat_docker_1.HardhatDocker.isInstalled())) {
        throw (0, utils_1.pluginError)('Docker Desktop is not installed.\n' +
            'Please install it by following the instructions on https://www.docker.com/get-started');
    }
}
exports.validateDockerIsInstalled = validateDockerIsInstalled;
async function createDocker() {
    return await handleCommonErrors(hardhat_docker_1.HardhatDocker.create());
}
exports.createDocker = createDocker;
async function pullImageIfNecessary(docker, image) {
    await handleCommonErrors(pullImageIfNecessaryInner(docker, image));
}
exports.pullImageIfNecessary = pullImageIfNecessary;
async function pullImageIfNecessaryInner(docker, image) {
    if (!(await docker.hasPulledImage(image))) {
        console.log(`Pulling Docker image ${hardhat_docker_1.HardhatDocker.imageToRepoTag(image)}...`);
        await docker.pullImage(image);
        console.log(`Image pulled`);
    }
    else {
        await checkForImageUpdates(docker, image);
    }
}
async function checkForImageUpdates(docker, image) {
    if (!(await docker.isImageUpToDate(image))) {
        console.log(`Updating Docker image ${hardhat_docker_1.HardhatDocker.imageToRepoTag(image)}...`);
        await docker.pullImage(image);
        console.log(`Image updated`);
    }
}
async function compileWithDocker(input, docker, image, config) {
    var _a, _b;
    const command = ['zksolc', '--standard-json'];
    if ((_b = (_a = config === null || config === void 0 ? void 0 : config.settings) === null || _a === void 0 ? void 0 : _a.optimizer) === null || _b === void 0 ? void 0 : _b.enabled) {
        command.push('--optimize');
    }
    // @ts-ignore
    const dockerInstance = docker._docker;
    return await handleCommonErrors(runZksolcContainer(dockerInstance, image, command, JSON.stringify(input)));
}
exports.compileWithDocker = compileWithDocker;
async function handleCommonErrors(promise) {
    try {
        return await promise;
    }
    catch (error) {
        if (error instanceof hardhat_docker_1.DockerNotRunningError || error instanceof hardhat_docker_1.DockerBadGatewayError) {
            throw (0, utils_1.pluginError)('Docker Desktop is not running.\nPlease open it and wait until it finishes booting.', error);
        }
        if (error instanceof hardhat_docker_1.DockerHubConnectionError) {
            throw (0, utils_1.pluginError)('Error connecting to Docker Hub.\nPlease check your internet connection.', error);
        }
        if (error instanceof hardhat_docker_1.DockerServerError) {
            throw (0, utils_1.pluginError)('Docker error', error);
        }
        if (error instanceof hardhat_docker_1.ImageDoesntExistError) {
            throw (0, utils_1.pluginError)(`Docker image ${hardhat_docker_1.HardhatDocker.imageToRepoTag(error.image)} doesn't exist.\n` +
                'Make sure you chose a valid zksolc version.');
        }
        throw error;
    }
}
